src/infer_request.cc

```diff
@@ -1269,47 +1274,89 @@ InferenceRequest::ValidateBytesInputs(
-                input_id + "', expecting " + std::to_string(element_count));
+  size_t remaining_element_size = 0;
+
+  size_t buffer_idx = 0;
+  const size_t buffer_count = input.Data()->BufferCount();
+
+  const char* buffer = nullptr;
+  size_t remaining_buffer_size = 0;
+  TRITONSERVER_MemoryType buffer_memory_type;
+  int64_t buffer_memory_id;
+
+  // Validate elements until all buffers have been fully processed.
+  while (remaining_buffer_size || buffer_idx < buffer_count) {
+    // Get the next buffer if not currently processing one.
+    if (!remaining_buffer_size) {
+      // Reset remaining buffer size and pointers for next buffer.
+      buffer = input.Data()->BufferAt(
+          buffer_idx++, &remaining_buffer_size, &buffer_memory_type,
+          &buffer_memory_id);
+      *expected_byte_size += remaining_buffer_size;
+
+      // FIXME: Skip GPU buffers for now, return an expected_byte_size of -1 as
+      // a signal to skip validation.
+      if (buffer_memory_type == TRITONSERVER_MEMORY_GPU) {
+        LOG_WARNING << LogRequest()
+                    << "Validation of GPU byte size buffers is not implemented "
+                       "yet, skipping validation for input: "
+                    << input_id;
+        *expected_byte_size = -1;
+        return Status::Success;
       }
+    }
 
-      const uint32_t len = *(reinterpret_cast<const uint32_t*>(content));
-      content += sizeof(uint32_t);
-      content_byte_size -= sizeof(uint32_t);
-      *expected_byte_size += sizeof(uint32_t);
-
-      if (content_byte_size < len) {
+    constexpr size_t kElementSizeIndicator = sizeof(uint32_t);
+    // Get the next element if not currently processing one.
+    if (!remaining_element_size) {
+      // FIXME: Assume the string element's byte size indicator is not spread
+      // across buffer boundaries for simplicity.
+      if (remaining_buffer_size < kElementSizeIndicator) {
         return Status(
             Status::Code::INVALID_ARG,
-            LogRequest() + "incomplete string data for inference input '" +
-                input_id + "', expecting string of length " +
-                std::to_string(len) + " but only " +
-                std::to_string(content_byte_size) + " bytes available");
+            LogRequest() +
+                "Element byte size indicator exceeds the end of the buffer.");
       }
 
-      content += len;
-      content_byte_size -= len;
-      *expected_byte_size += len;
+      // Start the next element and reset the remaining element size.
+      remaining_element_size = *(reinterpret_cast<const uint32_t*>(buffer));
       element_idx++;
+
+      // Advance pointer and remainder by the indicator size.
+      buffer += kElementSizeIndicator;
+      remaining_buffer_size -= kElementSizeIndicator;
+    }
+
+    // If the remaining buffer fits it: consume the rest of the element, proceed
+    // to the next element.
+    if (remaining_buffer_size >= remaining_element_size) {
```

Comment:
Please revisit the logic in this case. I think it will cause segfault/infinite loop if `remaining_buffer_size == remaining_element_size`.